source/analysis/AM1DDerivativeAB.cpp

```diff
@@ -153,10 +161,73 @@ void AM1DDerivativeAB::setInputSource()
 
 	reviewState();
 
-	emitSizeChanged(0);
+    emitSizeChanged();
 	emitValuesChanged();
-	emitAxisInfoChanged(0);
-	emitInfoChanged();
+    emitAxisInfoChanged();
+    emitInfoChanged();
+}
+
+void AM1DDerivativeAB::computeCachedValues() const
+{
+    AMnDIndex start = AMnDIndex(0);
+    AMnDIndex end = size()-1;
+    int totalSize = start.totalPointsTo(end);
+
+    QVector<double> data = QVector<double>(totalSize);
+    QVector<double> axis = QVector<double>(totalSize);
+    AMAxisInfo axisInfo = inputSource_->axisInfoAt(0);
+
+    inputSource_->values(start, end, data.data());
+
+    // This is much faster because we can compute all the axis values ourselves rather than ask for them one at a time.
+    if (axisInfo.isUniform){
+
+        double axisStart = double(axisInfo.start);
+        double axisStep = double(axisInfo.increment);
+
+        for (int i = 0; i < totalSize; i++)
+            axis[i] = axisStart + i*axisStep;
+
+        // Because we computed the axis values we are guarenteed that the values won't be bad.
+        cachedData_[0] = (data.at(1)-data.at(0))/(axis.at(1)-axis.at(0));
+        cachedData_[totalSize-1] = (data.at(totalSize-1)-data.at(totalSize-2))/(axis.at(totalSize-1)-axis.at(totalSize-2));
+
+        for (int i = 1, count = totalSize-1; i < count; i++)
+            cachedData_[i] = (data.at(i+1)-data.at(i-1))/(2*(axis.at(i+1)-axis.at(i-1)));
+    }
+
+    else {
+
+        // Fill the axis vector.  Should minimize the overhead of making the same function calls and casting the values multiple times.
+        for (int i = 0; i < totalSize; i++)
+            axis[i] = double(inputSource_->axisValue(0, i));
+
+        // Fill a list of all the indices that will cause division by zero.
+        QList<int> badIndices;
+
+        if (axis.at(0) == axis.at(1))
+            badIndices.append(0);
+
+        for (int i = 1, count = totalSize-1; i < count; i++)
+            if (axis.at(i+1) == axis.at(i-1))
+                badIndices.append(i);
+
+        if (axis.at(totalSize-1) == axis.at(totalSize-2))
+            badIndices.append(totalSize-1);
+
+        // Compute all the values
+        cachedData_[0] = (data.at(1)-data.at(0))/(axis.at(1)-axis.at(0));
+        cachedData_[totalSize-1] = (data.at(totalSize-1)-data.at(totalSize-2))/(axis.at(totalSize-1)-axis.at(totalSize-2));
+
+        for (int i = 1, count = totalSize-1; i < count; i++)
+            cachedData_[i] = (data.at(i+1)-data.at(i-1))/(2*(axis.at(i+1)-axis.at(i-1)));
```

Comment:
My calculus is pretty rusty, but isn't this making a note of the bad indices, and then ignoring them when it actually comes to do the calculation for a given ```i```?

I feel like this should be:
```C++
for (int i = 1, count = totalSize-1; i < count; i++) {
    if(!badIndices.contains(i))
        cachedData_[i] = (data.at(i+1)-data.at(i-1))/(2*(axis.at(i+1)-axis.at(i-1)));
}
```
Then the following loop would go through and fill in all the values which were missed because of bad indices. That said I would imagine the ```contains()``` method is probably O(numberOfBadIndices), and the chances of two values being a precise double match is slim - in which case maybe the badIndices should just be manually set as 0 and size -1, and then manually added?